src/test/java/integration/harness/MockSCMController.java

```diff
@@ -0,0 +1,498 @@
+        Repository repo = resolve(repository);
+        String hash = resolve(repository, branch).getHash();
+        Integer crNum = ++repo.lastChangeRequest;
+        repo.changes.put(crNum, hash);
+        repo.changeBaselines.put(crNum, branch);
+        return crNum;
+    }
+
+    public synchronized void closeChangeRequest(String repository, Integer crNum) throws IOException {
+        resolve(repository).changes.remove(crNum);
+        resolve(repository).changeBaselines.remove(crNum);
+    }
+
+    public synchronized String getTarget(String repository, Integer crNum) throws IOException {
+        return resolve(repository).changeBaselines.get(crNum);
+    }
+
+    public synchronized List<String> listBranches(String repository) throws IOException {
+        return new ArrayList<>(resolve(repository).heads.keySet());
+    }
+
+    public synchronized List<String> listTags(String repository) throws IOException {
+        return new ArrayList<>(resolve(repository).tags.keySet());
+    }
+
+    public synchronized List<Integer> listChangeRequests(String repository) throws IOException {
+        return new ArrayList<>(resolve(repository).changes.keySet());
+    }
+
+    public synchronized String getRevision(String repository, String branch) throws IOException {
+        return resolve(repository, branch).getHash();
+    }
+
+    public synchronized void addFile(String repository, String branchOrCR, String message, String path, byte[] content)
+            throws IOException {
+        Repository repo = resolve(repository);
+        String branchName;
+        Integer crNum;
+        String hash;
+        // check branch first
+        hash = repo.heads.get(branchOrCR);
+        if (hash == null) {
+            branchName = null;
+            Matcher m = Pattern.compile("change-request/(\\d+)").matcher(branchOrCR);
+            if (m.matches()) {
+                crNum = Integer.valueOf(m.group(1));
+                hash = repo.changes.get(crNum);
+                if (hash == null) {
+                    throw new IOException("Unknown change request: " + crNum + " in repository " + repository);
+                }
+            } else {
+                throw new IOException("Unknown branch: " + branchOrCR + " in repository " + repository);
+            }
+        } else {
+            branchName = branchOrCR;
+            crNum = null;
+        }
+        State base = repo.revisions.get(hash);
+        State state = new State(base, message, Collections.singletonMap(path, content), Collections.<String>emptySet());
+        repo.revisions.put(state.getHash(), state);
+        if (branchName != null) {
+            repo.heads.put(branchName, state.getHash());
+        }
+        if (crNum != null) {
+            repo.changes.put(crNum, state.getHash());
+        }
+    }
+
+    public synchronized void rmFile(String repository, String branchOrCR, String message, String path)
```

Comment:
This kind of stuff seems like it would be useful elsewhere; `workflow-multibranch` and similar currently use real SCM sources (for example with `GitSampleRepoRule`), but having a mock is sometimes nicer. Do you plan to

```xml
<no-test-jar>false</no-test-jar>
```

?